### ISSUE TITLE
Mesh Scale DLNR

### ISSUE THREAD

**Issue by Lk13Antkoo** [openvdb/fvdb-reality-capture#291]
Dear fVDB team,

I reconstructed the 3DGS, then I followed https://fvdb-reality-capture.readthedocs.io/latest/tutorials/radiance_field_and_mesh_reconstruction.html to reconstruct the mesh using DLNR :
v, f, c = mesh_from_splats_dlnr(model, cleaned_sfm_scene.camera_to_world_matrices, cleaned_sfm_scene.projection_matrices, cleaned_sfm_scene.image_sizes, truncation_margin)
u.save_mesh_vfc("reconstructed_mesh_dlnr.ply", v.cpu().numpy(), f.cpu().numpy(), c.cpu().numpy())

However, the output of the meshed is scaled between [0, 1], the Gaussian Splat scale does not. There is mismatch of scale between those presentation.

Could you please let me know how can I align those scales?

Thank you and have a nice day,
AnPham

**Comment by harrism** [comment]
Hi @Lk13Antkoo , thanks for the issue. 

`mesh_from_splats_dlnr` (and `mesh_from_splats`) return vertices in the coordinate frame of the `camera_to_world_matrices` / `projection_matrices` you pass in. The vertices come straight from `Grid.marching_cubes` on a TSDF grid built by unprojecting depths with those matrices — no separate normalization is applied. So the mesh lands in the same frame as `cleaned_sfm_scene`, which is the same frame the model was trained in.

So in the tutorial as written, the mesh and model _do_ share a frame. A scale mismatch means the splat you are looking at is in a different frame than the `cleaned_sfm_scene` cameras you passed to the mesher. The tutorial pipeline sets this up:

- `NormalizeScene(normalization_type="pca")` puts `cleaned_sfm_scene` — and thus model and the mesh — into a normalized training frame. (The transform is stored: `cleaned_sfm_scene.transformation_matrix`, world→normalized, also surfaced as `runner.reconstruction_metadata["normalization_transform"]`.)
- `model.save_ply(...)` saves means in that normalized frame (it does not bake in the transform — it just embeds the matrix as metadata).
- So the mismatch appears when the compared splat is in original/world coordinates: e.g. an external viewer that reads the PLY's `normalization_transform` metadata and applies it, a splat exported/handled through a world-frame path, or comparing against the original pre-normalization capture.

**How to align**

Option A — keep everything in the normalized/training frame (simplest). Save the splat with `model.save_ply("model.ply")` and the mesh with `pcu.save_mesh_vfc("mesh.ply", v…)`. Both are in the `cleaned_sfm_scene` frame and will overlap.

Option B — put the mesh into world/original coordinates (to match a world-frame splat):

```
import numpy as np
T = cleaned_sfm_scene.transformation_matrix          # 4x4, world -> normalized
Tinv = np.linalg.inv(T)                              # normalized -> world
v_np = v.cpu().numpy()
v_world = v_np @ Tinv[:3, :3].T + Tinv[:3, 3]        # affine: handles rotation+translation (+scale if present)
pcu.save_mesh_vfc("reconstructed_mesh_dlnr_world.ply", v_world, f.cpu().numpy(), c.cpu().numpy())
```

The affine form works whether or not the normalization includes scale (PCA is rigid; the similarity/geo normalizations do rescale).

(Claude helped create this answer.)

**Comment by harrism** [comment]
Here's a self-contained verification script (fvdb 0.5.0 / torch 2.11, single GPU — no dataset or DLNR weights needed):

```
# Empirical verification for fvdb-reality-capture issue #291 ("Mesh Scale DLNR").
#
# Claim verified:
#   * mesh_from_splats* returns vertices in the SAME coordinate frame as the camera
#     matrices / model you pass in.
#   * Comparing a normalized-frame mesh against world-scale splats reproduces the
#     reported scale mismatch.
#   * Applying inv(normalization_transform) to the mesh vertices re-aligns it to world.
#
# Mirrors reality-capture's tsdf_from_splats (batched / post-#292) using a synthetic
# Gaussian "bumpy plane" so no dataset/DLNR weights are needed.
import numpy as np
import torch
import fvdb
from fvdb import CameraModel, GaussianSplat3d, Grid

torch.manual_seed(0)
dev = "cuda"
dtype = torch.float16
print(f"fvdb {fvdb.__version__} | torch {torch.__version__} | cuda avail {torch.cuda.is_available()}")

# ----------------------------------------------------------------------------------
# Synthetic WORLD-scale scene: a bumpy plane of Gaussians at z~120, spanning [-40, 40]
# (a large, world-scale scene like the tutorial's +/-180).
# ----------------------------------------------------------------------------------
n = 48
xs = torch.linspace(-40, 40, n, device=dev)
gy, gx = torch.meshgrid(xs, xs, indexing="ij")
gz = 120.0 + 6.0 * torch.sin(gx * 0.12) * torch.sin(gy * 0.12)
means_w = torch.stack([gx.reshape(-1), gy.reshape(-1), gz.reshape(-1)], dim=1)
N = means_w.shape[0]
quats = torch.zeros(N, 4, device=dev); quats[:, 0] = 1.0
log_scales_w = torch.full((N, 3), float(np.log(1.5)), device=dev)
logit_opac = torch.full((N,), 8.0, device=dev)
sh0 = torch.rand(N, 1, 3, device=dev)
shN = torch.zeros(N, 0, 3, device=dev)

def make_model(means, log_scales):
    return GaussianSplat3d.from_tensors(means, quats, log_scales, logit_opac, sh0, shN)

model_w = make_model(means_w, log_scales_w)

# Cameras (world): identity rotation (look +z), a few x-translations for coverage.
H = W = 400
fx = fy = 500.0
proj = torch.tensor([[fx, 0, W / 2], [0, fy, H / 2], [0, 0, 1]], dtype=torch.float32, device=dev)
cam_x = [-20.0, 0.0, 20.0]
def cams_from_positions(positions):
    mats = []
    for px in positions:
        c = torch.eye(4, dtype=torch.float32, device=dev)
        c[:3, 3] = torch.tensor([px, 0.0, 0.0], device=dev)
        mats.append(c)
    return torch.stack(mats, 0)
c2w_w = cams_from_positions(cam_x)
proj_batch = proj.unsqueeze(0).repeat(len(cam_x), 1, 1)

# ----------------------------------------------------------------------------------
# Normalization transform T (world -> normalized), WITH scale (like NormalizeScene
# would apply for a scene needing rescaling): center at (0,0,120), uniform s = 1/80.
# ----------------------------------------------------------------------------------
s = 1.0 / 80.0
center = torch.tensor([0.0, 0.0, 120.0], device=dev)
T = torch.eye(4, device=dev)
T[0, 0] = T[1, 1] = T[2, 2] = s
T[:3, 3] = -s * center                      # p' = s*(p - center)

def apply_T(pts, M):
    return pts @ M[:3, :3].T + M[:3, 3]

# Build the NORMALIZED model + cameras (what training on a normalized SfmScene gives)
means_n = apply_T(means_w, T)
log_scales_n = log_scales_w + float(np.log(s))          # uniform scale s
model_n = make_model(means_n, log_scales_n)
c2w_n = c2w_w.clone()
for i in range(c2w_n.shape[0]):
    c2w_n[i, :3, 3] = apply_T(c2w_w[i, :3, 3].unsqueeze(0), T).squeeze(0)   # R = I here
# projection (intrinsics) are UNCHANGED by a world-space similarity -- same as reality-capture

# ----------------------------------------------------------------------------------
# tsdf_from_splats, batched (post-#292), returning marching-cubes vertices.
# ----------------------------------------------------------------------------------
@torch.no_grad()
def tsdf_mesh(model, c2w, projs, margin, shell=3.0, near=1e-4, far=1e10):
    device = model.device
    voxel = margin / shell
    grid = Grid.from_zero_voxels(voxel_size=voxel, origin=0.0, device=device)
    tsdf = torch.zeros(grid.num_voxels, device=device, dtype=dtype)
    weights = torch.zeros(grid.num_voxels, device=device, dtype=dtype)
    feats = torch.zeros((grid.num_voxels, 3), device=device, dtype=torch.uint8)
    for i in range(c2w.shape[0]):
        c = c2w[i].to(torch.float32)
        w2c = torch.linalg.inv(c).contiguous()
        p = projs[i].to(torch.float32)
        fd, alpha = model.render_images_and_depths(
            world_to_camera_matrices=w2c.unsqueeze(0), projection_matrices=p.unsqueeze(0),
            image_width=W, image_height=H, near=0.0, far=1e10, camera_model=CameraModel.PINHOLE)
        fimg = (fd[..., :3].clip_(0, 1) * 255).to(torch.uint8).squeeze(0)
        a = alpha[0].clamp(min=1e-10).squeeze(-1)
        depth = (fd[0, ..., -1] / a).to(dtype)
        wimg = ((depth > near) & (depth < far) & (a > 0.1)).to(dtype).squeeze(0)
        grid, tsdf, weights, feats = grid.integrate_tsdf_with_features(
            margin, p.to(dtype).unsqueeze(0), c.to(dtype).unsqueeze(0),
            tsdf, feats, weights, depth.unsqueeze(0), fimg.unsqueeze(0), wimg.unsqueeze(0))
    g2 = grid.pruned_grid(weights > 0.0)
    tsdf2 = g2.inject_from(grid, tsdf)
    v, f, _ = g2.marching_cubes(tsdf2, 0.0)
    return v.to(torch.float32)

# Reconstruct in each frame (margins scale with the frame)
v_world = tsdf_mesh(model_w, c2w_w, proj_batch, margin=3.0)          # world frame
v_norm  = tsdf_mesh(model_n, c2w_n, proj_batch, margin=3.0 * s)       # normalized frame
print(f"\nmesh vertex counts: world={v_world.shape[0]}, normalized={v_norm.shape[0]}")

# ----------------------------------------------------------------------------------
# Metrics
# ----------------------------------------------------------------------------------
def bbox(p):
    lo = p.min(0).values.cpu().numpy(); hi = p.max(0).values.cpu().numpy()
    return lo, hi, (hi - lo)

def nn_err(a, b, k=3000):
    # one-way mean nearest-neighbor distance from a to b, relative to bbox diagonal of b
    if a.shape[0] > k:
        a = a[torch.randperm(a.shape[0], device=a.device)[:k]]
    d = torch.cdist(a, b).min(dim=1).values.mean().item()
    _, _, ext = bbox(b)
    return d, d / float(np.linalg.norm(ext))

def show(name, p):
    lo, hi, ext = bbox(p)
    print(f"  {name:34s} bbox min={np.round(lo,3)} max={np.round(hi,3)} extent={np.round(ext,3)}")

print("\n--- Bounding boxes ---")
show("WORLD gaussian means", means_w)
show("mesh (world frame)", v_world)
show("NORMALIZED gaussian means", means_n)
show("mesh (normalized frame)", v_norm)

# (1) mesh is in the same frame as the model you passed:
_, r_world = nn_err(v_world, means_w)
_, r_norm  = nn_err(v_norm,  means_n)
print("\n(1) mesh lands ON the gaussian surface in its OWN frame (rel. NN err):")
print(f"      world mesh vs world means = {r_world:.4f}")
print(f"      norm  mesh vs norm  means = {r_norm:.4f}")

# (2) the symptom: normalized mesh vs WORLD splats -> mismatched by the scale factor
_, _, ext_world = bbox(means_w)
_, _, ext_norm_mesh = bbox(v_norm)
ratio = np.linalg.norm(ext_world) / np.linalg.norm(ext_norm_mesh)
print(f"\n(2) SYMPTOM: world-splat extent / normalized-mesh extent = {ratio:.1f}x  (expected ~{1/s:.0f}x)")

# (3) the fix: apply inv(T) to the normalized mesh -> should re-align to WORLD frame
Tinv = torch.linalg.inv(T)
v_norm_to_world = apply_T(v_norm, Tinv)                 # <-- the exact recipe from the reply
show("inv(T) @ mesh(norm)", v_norm_to_world)
_, r_realign_means = nn_err(v_norm_to_world, means_w)
_, r_realign_mesh  = nn_err(v_norm_to_world, v_world)
print(f"\n(3) FIX: inv(T) @ normalized-mesh vs WORLD means = {r_realign_means:.4f}")
print(f"         inv(T) @ normalized-mesh vs WORLD mesh  = {r_realign_mesh:.4f}")

ok = (r_norm < 0.02 and r_world < 0.02 and ratio > 40 and r_realign_means < 0.02)
print("\nRESULT:", "VERIFIED (mesh is in the passed-in frame; inv(T) re-aligns it)" if ok
      else "inconclusive -- inspect numbers above")
```

Output it produced:

mesh vertex counts: world=11706, normalized=11712

--- Bounding boxes ---
  WORLD gaussian means        bbox min=[-40. -40. 114.009] max=[40. 40. 125.991] extent=[80. 80. 11.982]
  mesh (world frame)          bbox min=[-44. -44. 114.]    max=[43.031 43.25 125.938] extent=[87.031 87.25 11.938]
  NORMALIZED gaussian means   bbox min=[-0.5 -0.5 -0.075]  max=[0.5 0.5 0.075] extent=[1. 1. 0.15]
  mesh (normalized frame)     bbox min=[-0.55 -0.55 -0.075] max=[0.538 0.541 0.074] extent=[1.088 1.091 0.149]

(1) world mesh vs world means = 0.0105
    norm  mesh vs norm  means = 0.0106
(2) SYMPTOM: world-splat extent / normalized-mesh extent = 73.5x  (expected ~80x)
(3) inv(T) @ normalized-mesh vs WORLD means = 0.0105
    inv(T) @ normalized-mesh vs WORLD mesh  = 0.0002

RESULT: VERIFIED (mesh is in the passed-in frame; inv(T) re-aligns it)

I'm going to go ahead and close this issue, but feel free to reopen if you feel it hasn't been addressed.
